src/sql/src/pure.rs

```diff
@@ -813,7 +626,223 @@ pub async fn purify_create_source(
 
     purify_source_format(&*catalog, format, connection, envelope, &connection_context).await?;
 
-    Ok((subsources, stmt))
+    Ok((subsources, Statement::CreateSource(stmt)))
+}
+
+async fn purify_alter_source(
+    catalog: Box<dyn SessionCatalog>,
+    mut stmt: AlterSourceStatement<Aug>,
+    connection_context: ConnectionContext,
+) -> Result<
+    (
+        Vec<(GlobalId, CreateSubsourceStatement<Aug>)>,
+        Statement<Aug>,
+    ),
+    PlanError,
+> {
+    let scx = StatementContext::new(None, &*catalog);
+    let AlterSourceStatement {
+        source_name,
+        action,
+        if_exists,
+    } = &mut stmt;
+
+    // Get connection
+    let pg_source_connection = {
+        // Get name.
+        let item = match scx.resolve_item(RawItemName::Name(source_name.clone())) {
+            Ok(item) => item,
+            Err(_) if *if_exists => {
+                return Ok((vec![], Statement::AlterSource(stmt)));
+            }
+            Err(e) => return Err(e),
+        };
+
+        // Ensure it's an ingestion-based and alterable source.
+        let desc = match item.source_desc()? {
+            Some(desc) => desc,
+            None => {
+                sql_bail!("cannot ALTER this type of source")
+            }
+        };
+
+        if !matches!(action, AlterSourceAction::AddSubsources { .. }) {
+            return Ok((vec![], Statement::AlterSource(stmt)));
+        }
+
+        match &desc.connection {
+            GenericSourceConnection::Postgres(pg_connection) => pg_connection.clone(),
+            _ => sql_bail!(
+                "{} is a {} source, which does not support ALTER TABLE...ADD SUBSOURCES",
+                scx.catalog.minimal_qualification(item.name()),
+                desc.connection.name()
+            ),
+        }
+    };
+
+    // If there's no further work to do here, early return.
```

Comment:
Is this comment accurate? It doesn't seem like there's an early return here.

Reply:
Code drifted and this comment didn't follow––fixed